controllers/metal3.io/preprovisioningimage_controller.go

```diff
@@ -90,56 +117,102 @@ func (r *PreprovisioningImageReconciler) Reconcile(ctx context.Context, req ctrl
 	return result, err
 }
 
+func configChanged(img *metal3.PreprovisioningImage, format metal3.ImageFormat, networkDataStatus metal3.SecretStatus) bool {
+	return !(img.Status.Format == format &&
+		img.Status.Architecture == img.Spec.Architecture &&
+		img.Status.NetworkData == networkDataStatus)
+}
+
 func (r *PreprovisioningImageReconciler) update(img *metal3.PreprovisioningImage, log logr.Logger) (bool, error) {
 	generation := img.GetGeneration()
 
-	url, format, errorMessage := getImageURL(img.Spec.AcceptFormats)
-	if errorMessage != "" {
-		log.Info("no suitable image URL available", "preferredFormat", format)
-		return setError(generation, &img.Status, reasonImageConfigurationError, errorMessage), nil
+	if !r.ImageProvider.SupportsArchitecture(img.Spec.Architecture) {
+		log.Info("image architecture not supported", "architecture", img.Spec.Architecture)
+		return setError(generation, &img.Status, reasonImageConfigurationError, "Architecture not supported"), nil
+	}
+
+	format := r.getImageFormat(img.Spec, log)
+	if format == "" {
+		return setError(generation, &img.Status, reasonImageConfigurationError, "No acceptable image format supported"), nil
 	}
 
-	log.Info("image URL available", "url", url, "format", format)
 	secretManager := secretutils.NewSecretManager(log, r.Client, r.APIReader)
-	secretStatus, err := getNetworkDataStatus(secretManager, img)
-	if err == nil {
-		return setImage(generation, &img.Status, url, format,
-			secretStatus, img.Spec.Architecture,
-			"Set default image"), nil
+	networkData, secretStatus, err := getNetworkData(secretManager, img)
+	if err != nil {
+		if k8serrors.IsNotFound(err) {
+			log.Info("network data Secret does not exist")
+			return setError(generation, &img.Status, reasonImageMissingNetworkData, "NetworkData secret not found"), err
+		}
+		return false, err
 	}
 
-	if k8serrors.IsNotFound(err) {
-		log.Info("network data Secret does not exist")
-		return setError(generation, &img.Status, reasonImageMissingNetworkData, "NetworkData secret not found"), err
+	if configChanged(img, format, secretStatus) {
+		reason := "Config changed"
+		if meta.IsStatusConditionTrue(img.Status.Conditions, string(metal3.ConditionImageReady)) {
+			// Ensure we mark the status as not ready before we remove the build
+			// from the image cache.
+			setUnready(generation, &img.Status, reason)
+		} else {
```

Comment:
nit: shouldn't this code branch be executed in both cases, not only when the annotation is not present?

Reply:
No, see comment on ll. 152-153.
This ensures that we are able to write to the CR that the image is not ready before we remove it. If we removed it first and then writing to the CR fails, we could end up in a state where the CR claims the URL is valid but it is in fact not for a possibly extended period of time.